### DISCUSSION TITLE
Can't start nextPYP: java.nio.file.FileSystemException: /mnt/data0: Function not implemented

### DISCUSSION THREAD

**Discussion by asarnow** [nextpyp/.github#103] [Q&A] [answered]
I'm trying to update nextPYP on my standalone workstation, but I get this error when I try to start up the service. I've also tried reinstalling from scratch. I'm not sure where it's reading the problematic path, it doesn't seem to matter if I change the binds or scratch directory. (I also don't know why it's problematic, /mnt/data0 is just a ext4 filesystem on a large SSD, and I haven't changed anything since nextPYP was working last).

```
Using configuration at /mnt/data0/applications/nextpyp/config.toml
Reading config.toml using CLI tool ...
WARNING: /etc/singularity/ exists, migration to apptainer by system administrator is not complete
Exception in thread "main" java.lang.ExceptionInInitializerError
        at edu.duke.bartesaghi.micromon.linux.CudaKt.countCudaGpus(cuda.kt:20)
        at edu.duke.bartesaghi.micromon.Config$Standalone.<clinit>(Config.kt:137)
        at edu.duke.bartesaghi.micromon.Config.<init>(Config.kt:225)
        at edu.duke.bartesaghi.micromon.Config$Companion.fromCanon(Config.kt:34)
        at edu.duke.bartesaghi.micromon.CliKt.main(cli.kt:12)
Caused by: java.nio.file.FileSystemException: /mnt/data0: Function not implemented
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:100)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:397)
        at java.base/java.nio.file.Files.createDirectory(Files.java:700)
        at java.base/java.nio.file.Files.createAndCheckIsDirectory(Files.java:807)
        at java.base/java.nio.file.Files.createDirectories(Files.java:793)
        at edu.duke.bartesaghi.micromon.IoKt.createDirsIfNeeded(io.kt:190)
        at edu.duke.bartesaghi.micromon.Config.initDirs(Config.kt:280)
        at edu.duke.bartesaghi.micromon.Backend.<clinit>(Backend.kt:34)
        ... 5 more
```

**Comment by cuchaz** [comment]
Guessing from the information in that error, that `/mnt/data0` path is coming from either your `web.localDir` or `web.sharedDir` setting. For some reason, nextpyp isn't allowed to create a folder there. The underlying filesystem is returning a "Function not implemented" error. I don't know what could cause that particular filesystem error (maybe the mount is using a userspace filesystem driver, like something from FUSE land?), but maybe take a look at the folders in your web config and see if anything looks out of place there.

**Reply by asarnow** [reply]
It's just a ext4 FS mounted via an /etc/fstab entry, there is a different FUSE mount in the binds, but it's outside the `/mnt/data0` tree and removing it didn't make a difference. My `localDir` is `/mnt/data0/applications/nextpyp/local` and `sharedDir` is `/mnt/data0/applications/nextpyp/shared`, the nextpyp user can create a directory at every level there except `/mnt`.

This is an Ubunut 20.04 system and `PYP_USER=PYP_GROUP=nextpyp`.

**Reply by cuchaz** [reply]
If your `localDir` or your `sharedDir` doesn't exist, nextpyp will try to create them with essentially `mkdir -p`. It doesn't look like that's working for some reason. You can try just creating the folders yourself (assuming they don't already exist) and then nextpyp should skip this step entirely.

**Reply by asarnow** [reply]
What about the child directories like `local/db` and so forth? I did originally create local, shared, and their children, because I was copying the previous install. I tried removing local & shared, or creating local & shared but not their children, but so far I always get the same error.

**Reply by asarnow** [reply]
Based on the code for `java.nio.file.Files.createDirectory` I think it's trying to create `/mnt/data0`?

Here's my `config.toml` excluding comments and some whitespace.

```
[pyp]
container = '/mnt/data0/applications/nextpyp/pyp.sif'
scratch = '/mnt/data0/asarnow/pyp'
binds = [ '/mnt/data0', '/data', '/home/asarnow' ]

[standalone]
availableCpus = 64
availableGpus = 2

[web]
localDir = '/mnt/data0/applications/nextpyp/local'
sharedDir = '/mnt/data0/applications/nextpyp/shared'
```

And nextpyp's .bashrc has these lines:

```
export PATH=$PATH:/mnt/data0/applications/nextpyp
export PYP_CONFIG=/mnt/data0/applications/nextpyp/config.toml
```

**Reply by cuchaz** [reply]
The server process is failing because it's trying to make sure your local and shared directories exist, but not only is the filesystem telling the server process they don't exist, but also that the server process can't create them. So at this point, the subfolders don't matter. But it's weird that we're getting conflicting information from the filesystem.

I'm not sure how to debug a filesystem issue like this remotely. The only thing I can think to check is that apptainer is getting the filesystem inside the container right. Are there any strange warnings or errors in the apptainer part of the system startup log? [The system logs for the nextpyp daemon](https://nextpyp.app/files/pyp/latest/docs/install/troubleshooting.html#console-output) would be the place to look for that. If any, they might show up near a line that looks like `Starting singularity container ...`.

It might also be good to check the arguments to apptainer/singularity, [particularly this `$binds` variable in the startup script](https://github.com/nextpyp/next/blob/main/run/start#L222). Maybe you can echo that to the log?

**Reply by cuchaz** [reply]
oh, wait. I see the problem in your config... Bind nesting in the apptainer arguments would probably cause this issue. Get rid of that `/mnt/data0` bind and I bet the problem will go away.

**Reply by asarnow** [reply]
That doesn't work, just tested again to make sure. (For now I have `scratch = '/tmp'` and `binds = ['/tmp']`).

**Reply by asarnow** [reply]
Based on the [java](https://github.com/frohoff/jdk8u-jdk/blob/master/src/solaris/classes/sun/nio/fs/UnixFileSystemProvider.java#L372) I think the string in the exception is the directory that `mkdir` is called on (by java via the micromon backend, not the nextpyp-start bash script), which for some reason is just `/mnt/data0`.

**Reply by cuchaz** [reply]
Bummer. Maybe there are more conflicts in the apptainer binds then. I'd have to see the whole contents of that $binds variable to be sure. Having that `/mnt/data0` bind there would definitely conflict with the binds the startup script makes for the local and shared folders because `/mnt/data0` exactly prefix matches the other folders. Prefix matches like that tend to confuse apptainer which could cause all kinds of weird filesystem issues. Seeing the apptainer/singularity part of the startup log might give more clues too.

**Reply by asarnow** [reply]
It's these lines that are failing:

```
120: # get the local and share directories from the config
121: localdir=$($cli localdir)
122: shareddir=$($cli shareddir)
```

**Reply by cuchaz** [reply]
> Based on the [java](https://github.com/frohoff/jdk8u-jdk/blob/master/src/solaris/classes/sun/nio/fs/UnixFileSystemProvider.java#L372) I think the string in the exception is the directory that mkdir is called on (by java via the micromon backend, not the nextpyp-start bash script), which for some reason is just /mnt/data0.

Yes, that's how that mkdir -p-like part of java code works, but that code path will never be reached if the folder exists already. [See here, just a few stack frames higher.](https://github.com/nextpyp/next/blob/main/src/backendMain/kotlin/edu/duke/bartesaghi/micromon/io.kt#L189)

**Reply by asarnow** [reply]
I changed scratch, binds, and workflowDirs (just in case) to consist only of an empty directory `/home/nextpyp/sandbox`, no difference.

**Reply by cuchaz** [reply]
> It's these lines that are failing:

How are they failing?

Also, are there any apptainer/singularity errors in the system log? What's the value of the `$binds` variable in that startup script I linked before? I won't be able to help much more without this information.

**Reply by asarnow** [reply]
Btw I added an `echo $binds` at line 119 in the start script, it gives
`--bind /mnt/data0/applications/nextpyp/config.toml:/var/micromon/config.toml`
And `/var/micromon` does not exist.

**Reply by asarnow** [reply]
Also there's nothing else in the nextpyp log, that's not in the console output I've posted above.

**Reply by cuchaz** [reply]
That fine, it will exist inside the container. You're missing basically all the binds though. That would explain the issues you're getting. Nothing is visible inside the container. Something has gone horribly wrong with the startup script, but I can't tell what.

**Reply by cuchaz** [reply]
Oh wait, you might be echoing `$binds` too early. Try echoing it at the bottom of the script, just before the `singularity` command.

**Reply by cuchaz** [reply]
wait a minute... you're getting java exceptions in the systemd log?? Not the micromon log? That's very weird. I missed that detail earlier. We almost never get java exceptions there, so I was just assuming the exception was happening during the JVM server process startup. But you're actually not getting to the bottom of the startup script, right? The exception is happening inside of the `$($cli ...)` subshell? I'll have to think for a minute about what went wrong since that's a very different scenario than was I was thinking about before.

**Reply by cuchaz** [reply, marked as answer]
Wow, ok. Found the cause of the issue. It's complicated.

Turns out a bug in my code was causing the server startup routines to be called during the CLI processing of the config file, which is waaay too early. That's why it looked like the server process was starting, because it was actually trying to do that, but that's not the right time to do it. I've already fixed the bug in the code, but you'll have to wait for the next release to get the fix.

In the meantime, you can work around the issue by putting the local and shared folders somewhere under your home directory. Through sheer dumb luck, that takes a code path that avoids the bug. If you need more space than your home folder has, you should be able to symlink your `/mnt/data0` folder into your home folder and use that. eg. `/home/me/symlink-to-mnt-0data0/local`.

**Reply by asarnow** [reply]
Great, that was it! I'm using the symlink workaround for now. Thanks!

**Reply by asarnow** [reply]
I think this caused something bad to happen. There were some jobs running last night but this morning the home directory links are real, and the `/mnt/data0/.../shared` contains a new link called shared that points to itself (which ofc. is not allowed).

**Reply by asarnow** [reply]
Local too.
<img width="630" alt="image" src="https://github.com/nextpyp/.github/assets/4052606/f3ae8b7b-00a2-4b23-9274-1cd1bb84dec6">

**Reply by cuchaz** [reply]
Weird. I have no idea what would cause what you're seeing. But now I can't get the workaround working again either (different errors though). Maybe it was a fluke it worked the first time at all.

Maybe it's best to either wait for the next version that has the fix for this issue already, or temporarily downgrade back to the old version.

**Reply by asarnow** [reply]
The explanation is I made a mistake creating symlinks, so it was just writing in the home directory. That worked until the filesystem was full. It seems the symlink workaround doesn't actually work, now that I copied the contents of local and shared back to the big SSD and added valid symlinks in /home.

**Reply by abartesaghi** [reply]
This issue was fixed in version [0.6.5](https://nextpyp.app/files/pyp/0.6.5/docs/index.html).
